### ISSUE TITLE
Number of GPUs used for image quantization

### ISSUE THREAD

**Issue by ChangyaoTian** [zh460045050/VQGAN-LC#3]
Hi, thanks for your great work!

I would like to know the exact number of GPUs used during the image quantization stage. I see in your paper that you mentioned "utilizing 32 Nvidia V100 GPUs", but the training script in the repo seems to indicate 8 cards (`--nproc-per-node 8`). So, what is the exact number of GPUs and the corresponding batch size?

**Comment by zh460045050** [comment]
We used a total of 4 nodes, each equipped with 8 V100 GPUs. Thus, the total number of GPUs was 32. The batch size was set to 256, resulting in 8 batches per GPU.

**Comment by ChangyaoTian** [comment]
Got it, thanks!

**Comment by ChangyaoTian** [comment]
Could you provide all the exact hyperparameters used for reproducing the training process of VQGAN-LC on ImageNet-1k? Cuz some key params are mismatched between the paper and the example script, such as the number of GPUS, batch size, lr and epochs.

![image](https://github.com/zh460045050/VQGAN-LC/assets/97594817/3bdd1987-0621-41fa-bf0b-a0fabfb791de)


![image](https://github.com/zh460045050/VQGAN-LC/assets/97594817/51b0b7a5-6898-4b95-aab5-84f2c66c02ea)

Thanks!

**Comment by ChangyaoTian** [comment]
Another mismatch found in codebook_generation: the default `k` value in `codebook_generation/minibatch_kmeans_per_class.py` is [1000](https://github.com/zh460045050/VQGAN-LC/blob/e52de8bff8948f97a77e94aa7629b75fa052665f/codebook_generation/minibatch_kmeans_per_class.py#L12), which means the generated codebook size will be 1M (1k \* 1k) rather than 100,000. 

Does it need to be changed to 100 instead?

**Comment by zh460045050** [comment]
Thank you for reporting these issues. The learning rate is set to 5e-4. We trained the tokenizers for only 20 epochs for comparison purposes, and we think longer training times may contribute to higher performance. Additionally, the k value in "minibatch_kmeans_per_class.py" should be set to 100 to generate a 100K codebook.

**Comment by ChangyaoTian** [comment]
Hey, I have some more new issues w.r.t the GPT training reproduction: 
The transformer in GPT has `509.42M` trainable params rather than the claimed `404M` param in paper, is that correct?

![image](https://github.com/zh460045050/VQGAN-LC/assets/97594817/e4b12f4a-5f42-45d3-bd48-18dd223635b6)

**Comment by ChangyaoTian** [comment]
And for GPT testing, there is no file named `eval_generation_imagenet.py`, is it should be `eval_generation.py` instead?

**Comment by ChangyaoTian** [comment]
As for GPT testing, the evaluation gt dir [here](https://github.com/zh460045050/VQGAN-LC/blob/02705d11abed9eae2b0f85909421e0e8f8fd1ce6/vqgan-gpt-lc/eval_generation.py#L244) should be the `val` subdir rather than `train` ? 

![image](https://github.com/zh460045050/VQGAN-LC/assets/97594817/371961e7-99b4-4f6a-9bdc-a02d1c320e08)

**Comment by zh460045050** [comment]
Thank you for reporting these issues.

1. For the size of GPT, the parameters (404M) are reported for only the transformer layers. Thus, we believe 509.42M is correct when including the token embeddings.

2. "eval_generation_imagenet.py" should be "eval_generation.py" and we have corrected it in the README.

3. For GPT testing, we follow existing works to compare the FID between 50,000 generated images and the ImageNet training set. Thus, the evaluation ground truth directory is correct.

**Comment by ChangyaoTian** [comment]
Thanks for your reply. For GPT testing, as far as I know, nowadays a more common practise is to directly adopt the evaluation script from [OpenAI](https://github.com/openai/guided-diffusion/tree/main/evaluations), where the statistics are calculated over the whole dataset. Could you also provide the eval script and the corresponding results? Thanks!

**Comment by fpaissan** [comment]
Hey @ChangyaoTian (cc: @zh460045050). Where you able to reproduce the reconstruction results?

Using the provided command and script, I can only achieve the following results. 

| rFID ↓ | LPIPs ↓ | pSNR ↑ | SSIM ↑ | Codebook Util |
| :--- | :--- | :--- | :--- | :--- |
| 2.8508 | 0.1362 | 23.861 | 0.5954 | 99530/100K |

Any help would be appreciated.
